core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -366,25 +390,72 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
                 // A task should only be in the aborted state if the STAGE is done (ABORTED or FAILED)
                 fail(new TrinoException(GENERIC_INTERNAL_ERROR, "A task is in the ABORTED state but stage is " + stageState));
                 break;
-            case FLUSHING:
-                flushingTasks.add(taskStatus.getTaskId());
-                break;
-            case FINISHED:
-                finishedTasks.add(taskStatus.getTaskId());
-                flushingTasks.remove(taskStatus.getTaskId());
-                break;
             default:
+                break;
+        }
+    }
+
+    private void updateTaskStatus(TaskStatus taskStatus)
+    {
+        State stageState = stateMachine.getState();
+        TaskState taskState = taskStatus.getState();
+        if (stageState.isDone()) {
+            return;
+        }
+
+        if (taskState == TaskState.FAILED || taskState == TaskState.CANCELED || taskState == TaskState.ABORTED) {
+            updateNotSuccessfulTaskStatus(taskStatus);
+            return;
+        }
+
+        long flushingOrFinishedTasksCounter = 0;
+        long finishedTasksCounter = 0;
+
+        /*
+         * We are counting the unique number of flushing/finished tasks using the atomic counter flushingOrFinishedTasksCounter.
+         * When some thread detects that number of finished or flushing tasks is equal
+         * to allTasks.size() then it decides to move stateMachine's state to FLUSHING/FINISHED. To distinguish between
+         * FLUSHING or FINISHED we use `flushingTaskWasObserved` to mark that there was at least one flushing task.
+        */
+        if (taskState == TaskState.FLUSHING) {
+            boolean addedNow = flushingOrFinishedTasks.add(taskStatus.getTaskId());
+            if (addedNow) {
+                flushingTaskWasObserved = true;
+                flushingOrFinishedTasksCounter = this.flushingOrFinishedTasksCounter.incrementAndGet();
+            }
+        }
+
+        /*
+         * We are counting the unique number of finished tasks using the atomic counter finishedTasksCounter.
+         * When some thread detects that number of finished tasks is equal to allTasks.size() then it decides
+         * to move stateMachine's state to FINISHED.
+         */
+        else if (taskState == TaskState.FINISHED) {
+            boolean addedNow = flushingOrFinishedTasks.add(taskStatus.getTaskId());
```

Comment:
What if task moves directly to `finished` state? It seems it should atomically be added to both `flushingOrFinishedTasks` and `finishedTasks`.

If you add `finished` task first to `flushingOrFinishedTasks` and only later to `finishedTasks` then for a while task will be considered flushing even though it's in fact finished.

Reply:
> If you add finished task first to flushingOrFinishedTasks and only later to finishedTasks then for a while task will be considered flushing even though it's in fact finished.

No, the task will be considered as a flushing or finished task, what is true. 
If the task moves directly to `finished` state, we firstly increment the counter `flushingOrFinishedTasksCounter`. It is possible that now (after incrementing that value) the stage is going to be a `FLUSHING` but we move to that state only if there was a `FLUSHING` task (look at the `flushingTaskWasObserved`). So when there was at least one `flushing` task and we observe that all tasks are `flushing` or `finished` we can safely move stage's state to the `FLUSHING`

Reply:
Why do we need extra counter rather than checking if `flushingOrFinishedTasks` or `finishedTasks` is empty/non-empty or checking size of these sets?
Extra counter seems like another entity that you need to keep updating in semi-consistent way?

Reply:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html 
You cannot use `size/empty/non-empty` to the program control because:
`Bear in mind that the results of aggregate status methods including size, isEmpty, and containsValue are typically useful only when a map is not undergoing concurrent updates in other threads. Otherwise the results of these methods reflect transient states that may be adequate for monitoring or estimation purposes, but not for program control.
`
> Extra counter seems like another entity that you need to keep updating in semi-consistent way?
What do you mean by `semi-consistent`?

Reply:
> What do you mean by semi-consistent?

Now you have to keep updating `finishedTasksCounter`, `flushingOrFinishedTasksCounter`, `flushingOrFinishedTasksCounter`, `flushingOrFinishedTasks` and `finishedTasks`. 
This seems pretty complicated

Reply:
I agree. This is a price of being lock-free for that method. I could to come up with another idea to make it thread-safe without additional counters